pkg/addoninstance/utils.go

```diff
@@ -0,0 +1,42 @@
+package addoninstance
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+)
+
+var heartbeatTimeoutCondition metav1.Condition = metav1.Condition{
+	Type:    "addons.managed.openshift.io/Healthy",
+	Status:  "Unknown",
+	Reason:  "HeartbeatTimeout",
+	Message: "Addon failed to send heartbeat.",
+}
+
+func parseTargetNamespaceFromAddon(addon addonsv1alpha1.Addon) (string, error) {
+	var targetNamespace string
+	switch addon.Spec.Install.Type {
+	case addonsv1alpha1.OLMOwnNamespace:
+		if addon.Spec.Install.OLMOwnNamespace == nil ||
+			len(addon.Spec.Install.OLMOwnNamespace.Namespace) == 0 {
+			// invalid/missing configuration
+			return "", fmt.Errorf(".install.spec.olmOwmNamespace.namespace not found")
+		}
+		targetNamespace = addon.Spec.Install.OLMOwnNamespace.Namespace
+
+	case addonsv1alpha1.OLMAllNamespaces:
+		if addon.Spec.Install.OLMAllNamespaces == nil ||
+			len(addon.Spec.Install.OLMAllNamespaces.Namespace) == 0 {
+			// invalid/missing configuration
+			return "", fmt.Errorf(".install.spec.olmAllNamespaces.namespace not found")
+		}
+		targetNamespace = addon.Spec.Install.OLMAllNamespaces.Namespace
+	default:
+		// ideally, this should never happen
+		// but technically, it is possible to happen if validation webhook is turned off and CRD validation gets bypassed via the `--validate=false` argument
+		return "", fmt.Errorf("unsupported install type found: %s", addon.Spec.Install.Type)
+	}
+	return targetNamespace, nil
+}
```

Comment:
I was initially trying to do that only but realised that it wouldn't be possible to do so here or would be very un-clean to do so.

That's because, the `pkg/addoninstance` although currently is a part of this repo, but still it is acting like an absolutely different entity from the addon-operator. (same reason why `pkg/addoninstance` should be, and I guess would be, moved to a separate repo).

Now, the `parseAddonInstallConfig` method you're talking about, expects an `AddonReconciler` object as its receiver.

This means that de-duplicating the above code you mentioned with the help of `parseAddonInstallConfig` would require an empty/dummy `AddonReconciler` object to be created unnecessarily. And I am not sure if that would be suitable or graceful enough too considering the usage and reliance on `r.reportConfigurationError` method under the `parseAddonInstallConfig`.

So, it just seemed better to simply create a (slightly redundant-looking) parseTargetNamespaceFromAddon function under the blanket of `addoninstance` repo/package/client-side sdk which it can make use of solely without unnecessarily needing to import the addon-operator's API as a dependency and create a dummy `AddonReconciler` out of it just for the sake of making use of the `parseAddonInstallConfig` method it serves.